handlers/viber/viber.go

```diff
@@ -10,3 +31,386 @@ POST /handlers/viber_public/uuid?sig=sig
+		case "url":
+			text = viberMsg.Message.Media
+		case "location":
+			mediaURL = fmt.Sprintf("geo:%f,%f", viberMsg.Message.Location.Latitude, viberMsg.Message.Location.Longitude)
+		case "text":
+			text = viberMsg.Message.Text
+		default:
+			return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, fmt.Errorf("unknown message type: %s", messageType))
+		}
+
+		if text == "" && mediaURL == "" {
+			return nil, courier.WriteAndLogRequestError(ctx, w, r, channel, fmt.Errorf("missing text or media in message in request body"))
+		}
+
+		// build our msg
+		msg := h.Backend().NewIncomingMsg(channel, urn, text).WithExternalID(fmt.Sprintf("%d", viberMsg.MessageToken)).WithContactName(contactName)
+
+		if mediaURL != "" {
+			msg.WithAttachment(mediaURL)
+		}
+
+		// and finally queue our message
+		err = h.Backend().WriteMsg(ctx, msg)
+		if err != nil {
+			return nil, err
+		}
+
+		return []courier.Event{msg}, courier.WriteMsgSuccess(ctx, w, r, []courier.Msg{msg})
+	}
+
+	return nil, courier.WriteError(ctx, w, r, fmt.Errorf("not handled, unknown event: %s", event))
+}
+
+type viberMessage struct {
+	Event        string `json:"event" validate:"required"`
+	Timestamp    int64  `json:"timestamp" validate:"required"`
+	MessageToken int64  `json:"message_token" validate:"required"`
+	Sender       struct {
+		ID   string `json:"id"`
+		Name string `json:"name"`
+	} `json:"sender"`
+	User struct {
+		ID   string `json:"id"`
+		Name string `json:"name"`
+	} `json:"user"`
+	Message struct {
+		Text    string `json:"text"`
+		Media   string `json:"media"`
+		Contact struct {
+			Name        string `json:"name"`
+			PhoneNumber string `json:"phone_number"`
+		}
+		Location struct {
+			Latitude  float64 `json:"lat"`
+			Longitude float64 `json:"lon"`
+		}
+		Type         string `json:"type"`
+		TrackingData string `json:"tracking_data"`
+	} `json:"message"`
+}
+
+// see https://developers.viber.com/docs/api/rest-bot-api/#callbacks
+func (h *handler) validateSignature(channel courier.Channel, r *http.Request) error {
+	actual := r.Header.Get(viberSignatureHeader)
+	if actual == "" {
+		return fmt.Errorf("missing request signature")
+	}
+
+	confAuth := channel.ConfigForKey(courier.ConfigAuthToken, "")
+	authToken, isStr := confAuth.(string)
+	if !isStr || authToken == "" {
+		return fmt.Errorf("invalid or missing auth token in config")
+	}
+
+	// read our body
+	body, err := ioutil.ReadAll(r.Body)
+	r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+	expected, err := viberCalculateSignature(authToken, body)
+	if err != nil {
+		return err
+	}
+
+	// compare signatures in way that isn't sensitive to a timing attack
+	if !hmac.Equal(expected, []byte(actual)) {
+		return fmt.Errorf("invalid request signature")
+	}
+
+	return nil
+}
+
+func viberCalculateSignature(authToken string, contents []byte) ([]byte, error) {
+
+	var buffer bytes.Buffer
+	buffer.Write(contents)
```

Comment:
what purpose does buffer serve? can't we just write the contents directly below?